framework/include/utils/InputParameterWarehouse.h

```diff
@@ -176,6 +177,8 @@ class InputParameterWarehouse
   /// The factory is allowed to call addInputParameters and removeInputParameters.
   friend MooseObjectPtr Factory::create(
       const std::string &, const std::string &, const InputParameters &, THREAD_ID, bool);
+  friend std::shared_ptr<Action>
+  ActionFactory::create(const std::string &, const std::string &, InputParameters &);
```

Comment:
Why is this necessary?

Reply:
This is for calling `addInputParameters` function similar as `Factory::create`.